Tidy up solver env

gridbugs opened the pull request:
This makes it possible to configure the `make` and `jobs` opam variables through a workspace file and also to derive the `make` variable (the path to the `make` executable) from `$PATH` using the same machinery as the `%{exe:make}` pform macro. As the semantics of setting and reading build config variables differs from other types of opam variable I've taken this opportunity to refactor the solver env to reflect the different types of opam variable. This unifies the way variables are accessed which allows simplifications to the solver, also included in this change.

Also since the size of the solver environment is growing large, this change modifies the way that the environment is printed when the `--just-print-solver-env` flag is passed. The output is now more verbose and formats the variables as a nested bulleted list rather than a sexp.

gridbugs committed:
Tidy up solver env

Refactor of the `Solver_env` module:
- separate constants from system environment variables
- add human-readable pretty-printer
- unify variables with `Variable.t` type which simplifies the solver

Signed-off-by: Stephen Sherratt <stephen@sherra.tt>

gridbugs commented:
I still intend to update tests to ensure we can parse the new variables from the workspace config as thus far I've relied on deriving them from the current system. Opening this PR now for visibility and early feedback as it's quite a large change.

rgrinberg commented on src/dune_pkg/solver_env.mli:
```diff
@@ -1,84 +1,99 @@
-val default : t
+(** Look up a variable's value by name *)
+val get_by_variable_name :
+  t -> string -> (Variable_value.t, [ `Unknown_variable ]) result
```
Could you introduce a type for variable names?

gridbugs replied:
There is `Variable.t` inside solver_env.ml but it's not exposed. Think it would be clearer to use expose it and get rid of this helper function?

gridbugs replied:
We would still need to expose a way to convert strings (or I guess `OpamVariable.t`s) into the `Variable.t` type though. What's the purpose of avoiding raw strings? Is it just because the meaning of the string is not suggested by its type? Would replacing `string` in this api with `OpamVariable.t` be sufficient?

rgrinberg replied:
There's two problem with using raw strings:

1. It makes reading the code harder than it needs be.
2. It becomes possible to skip the validation that is often necessary for these various types. For example, I'm pretty sure that an opam variable cannot contain the `'\0'` character. In some cases where we didn't add such validation, we ended up having to introduce painful workaround down the roads (see the code library names and module names).

Leonidas-from-XIV replied:
I would suggest against exposing `OpamVariable.t` in our API, as this ties us a bit stronger to OPAM and if we decide to change it later it would make things harder.

rgrinberg commented on src/dune_pkg/solver_env.mli:
```diff
@@ -1,84 +1,99 @@
+    circular dependency. *)
+val with_default_values_for_unset_build_configs :
+     With_opt_build_config.t
+  -> which_make:(unit -> Path.t option Memo.t)
```
Why does this argument need to be delayed?

gridbugs replied:
We only need to call it if there is no `make` variable configured.

rgrinberg replied:
Should we make `%{make}` part of the solver environment at all? Selecting a package based on the concrete path of make sounds like a very bad idea to me. Should we just forbid it?

gridbugs replied:
Good point. I added them so that we could generate lockfiles with the build and install commands fully evaluated as a first step before supporting filters and substitutions in build and install commands but given the complexity added by these variables and the fact that packages _shouldn't_ be filtering based on `%{make}` or `%{jobs}` I'm inclined to remove them. Instead I'll generate lockfiles with unfiltered commands and convert variables like `%{make}` and `%{jobs}` into the appropriate pforms.

gridbugs commented:
Updated the `just-print-solver-env` test to set the `make` variable in its workspace configs. This turns out to be necessary for the test as otherwise it will print the path to `make` on the system running the test which will differ between different systems.

gridbugs commented:
There are some tests failing with:
```
[ERROR] The selected switch default is not installed.
'/opt/hostedtoolcache/opam/2.1.5/x86_64/opam config env --switch default --sexp --set-switch' failed.
[1]
```
This seems to be because we have to run that opam command to determine the location of `make` since it could be inside opam's `bin` dir. I'll debug this further tomorrow.

gridbugs commented:
Based on the fact that we really don't want packages to be able to filter based on the path to `make` or the `jobs` value I've removed the build config variables from this PR and changed the name accordingly. I've kept the solver env refactors as they simplify the solver and lock command implementation.

rgrinberg commented:
@gridbugs please rebase (and reformat)

gridbugs commented:
> @gridbugs please rebase (and reformat)

Done!
